File: hackerrank-java/src/com/company/ProductExceptSelf.java
package com.company;

import java.util.Arrays;

public class ProductExceptSelf {

    public static int[] productExceptSelf(int[] nums) {
        int[] prefixProduct = new int[nums.length + 2];
        int[] suffixProduct = new int[nums.length + 2];
        int[] result = new int[nums.length];
        prefixProduct[0] = 1;
        suffixProduct[nums.length + 1] = 1;


        // i  j
        // 6  3

        //   j
        //     1  2  3   4
        //  1  1  2  6   24  0
        //  0  24 24 12  4   1
        //     24, 12, 8, 6

        //           i
        // 1 1 2 6 24 1

        //     i
        // 1  0  0  0   0  1
        // 1  24 24 12  4  1


        for (int i=1; i<nums.length; i++) {
            prefixProduct[i] = prefixProduct[i-1] * nums[i-1];
        }

        for (int i=suffixProduct.length-2; i>0; i--) {
            suffixProduct[i] = suffixProduct[i+1] * nums[i-1];
        }



        for (int i=0; i<nums.length; i++) {
            result[i] = prefixProduct[i] * suffixProduct[i+2];

        }
        return result;

    }


    public static void main(String[] args) {
        System.out.println(Arrays.toString(productExceptSelf(new int[]{1,2,3,4})));
    }
}

File: product_except_self/product_except_self.go
package main

import "fmt"

func main() {
	nums := []int{1, 2, 3, 4}
	fmt.Printf("%v", productExceptSelf(nums))

}

func productExceptSelf(nums []int) []int {

	prefixProduct := make([]int, len(nums)+2)
	suffixProduct := make([]int, len(nums)+2)
	result := make([]int, len(nums))
	prefixProduct[0] = 1
	suffixProduct[len(nums)+1] = 1

	for i := 1; i < len(nums); i++ {
		prefixProduct[i] = prefixProduct[i-1] * nums[i-1]
	}

	for i := len(suffixProduct) - 2; i > 0; i-- {
		suffixProduct[i] = suffixProduct[i+1] * nums[i-1]
	}

	for i := 0; i < len(nums); i++ {
		result[i] = prefixProduct[i] * suffixProduct[i+2]

	}
	return result
}
